polkadot/node/core/dispute-coordinator/src/scraping/candidates.rs

```diff
@@ -135,11 +135,14 @@ impl ScrapedCandidates {
 	}
 
 	pub fn insert(&mut self, block_number: BlockNumber, candidate_hash: CandidateHash) {
-		self.candidates.insert(candidate_hash);
-		self.candidates_by_block_number
+		if self
+			.candidates_by_block_number
 			.entry(block_number)
 			.or_default()
-			.insert(candidate_hash);
+			.insert(candidate_hash)
+		{
+			self.candidates.insert(candidate_hash);
+		}
```

Comment:
How is this different from previous logic? I don't really understand, it should have worked for multiple blocks per height previously. This if will fail only in case of insertion of the same CandidateHash, which shouldn't happen, but even then it would be the same as previously.

Reply:
Oh, it's because of ref counting. nvm

Reply:
No, because the inserts in candidates is ref counted. Therefore if we insert at the same height multiple times, we would increase the refcount each time. But in cleanup we would only decrement it once, hence we would never clean it.

Demonstration: The modified test fails without that change (pruning does not happen).

In case of forks there can easily be the same candidate included twice at the same height.

Reply:
But this will only happen if we insert the same candidate (hash), not different candidates at the same height, no? How is this possible?

Reply:
I guess if there are multiple forks at the same height including the same candidate.

Reply:
Yes same hash, same height.

```
     - B
A<
     - B'
``` 
    
     
 A candidate gets backed in `A`, afterwards we get a fork. Bitfields will be imported in both forks. Result: same candidate included at the same height.

Reply:
This is really nasty. Good catch!